crates/matrix-sdk/src/event_cache/linked_chunk.rs

```diff
@@ -43,48 +38,156 @@ pub enum LinkedChunkError {
+    fn latest_chunk(&self) -> &Chunk<CAP, Item, Gap> {
+        unsafe { self.last.unwrap_or(self.first).as_ref() }
+    }
+
+    /// Get the latest chunk, as a mutable reference.
+    fn latest_chunk_mut(&mut self) -> &mut Chunk<CAP, Item, Gap> {
+        unsafe { self.last.as_mut().unwrap_or(&mut self.first).as_mut() }
+    }
+
+    /// Get the chunk as a reference, from its identifier, if it exists.
+    fn chunk(&self, identifier: ChunkIdentifier) -> Option<&Chunk<CAP, Item, Gap>> {
+        let mut chunk = self.latest_chunk();
+
+        loop {
+            if chunk.identifier() == identifier {
+                return Some(chunk);
+            }
+
+            chunk = chunk.previous()?;
+        }
+    }
+
+    /// Get the chunk as a mutable reference, from its identifier, if it exists.
+    fn chunk_mut(&mut self, identifier: ChunkIdentifier) -> Option<&mut Chunk<CAP, Item, Gap>> {
+        let mut chunk = self.latest_chunk_mut();
+
+        loop {
+            if chunk.identifier() == identifier {
+                return Some(chunk);
+            }
+
+            chunk = chunk.previous_mut()?;
+        }
+    }
+}
+
 /// The [`LinkedChunk`] structure.
 ///
 /// It is similar to a linked list, except that it contains many items `Item`
 /// instead of a single one. A chunk has a maximum capacity of `CHUNK_CAPACITY`.
 /// Once a chunk is full, a new chunk is created. Not all chunks are necessarily
 /// entirely full. A chunk can represents a `Gap` between other chunks.
-pub struct LinkedChunk<const CHUNK_CAPACITY: usize, Item, Gap, Listener = ()> {
-    /// The first chunk.
-    first: NonNull<Chunk<CHUNK_CAPACITY, Item, Gap>>,
-    /// The last chunk.
-    last: Option<NonNull<Chunk<CHUNK_CAPACITY, Item, Gap>>>,
+///
+/// ⚠️ When the `LinkedChunk` is manipulated, all updates (i.e. operations) are
+/// kept in memory. It's like a history of all `LinkedChunk`'s updates. This is
+/// useful when one wants to replicate the `LinkedChunk` in another kind of
+/// storage, like a database. Every time one calls `LinkedChunk::updates()` to
+/// read the updates, it's mandatory to clean, to drain, to empty it. Even if
+/// one doesn't care about the updates, it's necessary to empty the updates,
+/// otherwise it will grow in memory for the eternity.
+pub struct LinkedChunk<const CHUNK_CAPACITY: usize, Item, Gap> {
+    /// The links to the chunks, i.e. the first and the last chunk.
+    links: LinkedChunkLinks<CHUNK_CAPACITY, Item, Gap>,
     /// The number of items hold by this linked chunk.
     length: usize,
     /// The generator of chunk identifiers.
     chunk_identifier_generator: ChunkIdentifierGenerator,
-    /// Listener, i.e. a type that will receive updates from what's happening
-    /// inside the `LinkedChunk`.
-    listener: Arc<Listener>,
+    /// All updates that have been made on this `LinkedChunk`. Note that it can
+    /// be drained, so it doesn't contain all updates since it exists.
```

Comment:
nit: two `it` referencing two different things, I think, can you clarify the second `it` please?